### PR TITLE
-Support for Sink Codecs

### LINKED ISSUE #2403: Support a generic codec structure for sinks
**Is your feature request related to a problem? Please describe.**

Data Prepper supports a file sink and has plans to support an S3 Sink (#1048). Sinks like these can benefit from a sink codec concept similar to the source codec (#1532).

**Describe the solution you'd like**

Create an interface in `data-prepper-api` which can transform events into an output format for consumption by participating sinks.

### Proposed interface

```
interface OutputCodec {

  /**
   * Called by the sink when a new destination is in use.
   */
  void start(OutputStream outputStream);

  /**
   * Called by the sink before saving the destination to the external sink.
   */
  void complete(OutputStream outputStream);

  /**
   * Writes an event into the underlying stream.
   */
  void writeEvent(Event event, OutputStream outputStream);

  /**
   * Returns the expected file extension for this type of codec.
   * For example, json, csv.
   */
  String getExtension();
}
```

I included the `start` and `complete` interfaces because some codecs like JSON array need to create an initial wrapping and then close that. 

Additionally, the `getExtension` method can be used to determine the file name when the sink is writing to a file or file-like system (e.g. S3).

### Plugin names

Use the same name as the corresponding input plugin when appropriate. Thus, we will have an input plugin for `newline` and an output plugin for `newline`. The Data Prepper plugin framework permits this because they implement different interfaces.

### Project structure

Include these new output codecs in the same Gradle projects as the input codec. 

For example, both the input and output `newline` codecs should be in the `data-prepper-plugins/newline-codecs` project.

### PR DESCRIPTION
**Pull request by omkarmmore95**
Signed-off-by: omkarmmore95 <omkar.m.more95@gmail.com>

### Description
NDJson Codec refactor
 
### Issues Resolved
#2403 
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/parse-json-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/json/NewlineDelimitedOutputCodec.java [hunk 1]
```diff
@@ -19,7 +19,7 @@
  * An implementation of {@link OutputCodec} which deserializes Data-Prepper events
  * and writes them to Output Stream as ND-JSON data
  */
-@DataPrepperPlugin(name = "newline", pluginType = OutputCodec.class, pluginConfigurationType = NewlineDelimitedOutputConfig.class)
+@DataPrepperPlugin(name = "ndjson", pluginType = OutputCodec.class, pluginConfigurationType = NewlineDelimitedOutputConfig.class)
 public class NewlineDelimitedOutputCodec implements OutputCodec {
```
Comment (dlvenable) [inline data-prepper-plugins/parse-json-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/json/NewlineDelimitedOutputCodec.java:23]: Please also rename this class to `NdjsonOutputCodec`. Thanks!

File: data-prepper-plugins/parse-json-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/json/NewlineDelimitedOutputCodec.java [hunk 2]
```diff
@@ -19,7 +19,7 @@
  * An implementation of {@link OutputCodec} which deserializes Data-Prepper events
  * and writes them to Output Stream as ND-JSON data
  */
-@DataPrepperPlugin(name = "newline", pluginType = OutputCodec.class, pluginConfigurationType = NewlineDelimitedOutputConfig.class)
+@DataPrepperPlugin(name = "ndjson", pluginType = OutputCodec.class, pluginConfigurationType = NewlineDelimitedOutputConfig.class)
```
Comment (dlvenable) [inline data-prepper-plugins/parse-json-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/json/NewlineDelimitedOutputCodec.java:22]: Please rename `NewlineDelimitedOutputConfig` to `NdjsonOutputConfig`

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thanks for making this fix. I have a couple renaming comments.

File: data-prepper-plugins/parse-json-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/json/NewlineDelimitedOutputCodec.java (hunk 1 above)
Comment (omkarmmore95) [reply data-prepper-plugins/parse-json-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/json/NewlineDelimitedOutputCodec.java]: Hi @dlvenable , As suggested we have renamed the classes and all necessary changes have been implemented throughout the codebase.

File: data-prepper-plugins/parse-json-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/json/NewlineDelimitedOutputCodec.java (hunk 2 above)
Comment (omkarmmore95) [reply data-prepper-plugins/parse-json-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/json/NewlineDelimitedOutputCodec.java]: Hi @dlvenable , As suggested we have renamed the classes and all necessary changes have been implemented throughout the codebase.

**Review by kkondaka — APPROVED** [review]

**Pull request merged**